app/services/articles/feeds/weighted_query_strategy.rb

```diff
@@ -0,0 +1,576 @@
+    #
+    # Where this class differs is that it is aiming to build the feed
+    # based from the given user's perspective.  Whereas the other Feed
+    # algorithm starts with a list of candidates that are global to
+    # the given Forem (e.g., starting the base query from the
+    # `articles.score`, a volatile and swingy value that favors global
+    # reactions over user desired content).
+    #
+    # This is not quite a chronological only feed but could be easily
+    # modified to favor that.
+    #
+    # @note One possible short-coming is that the query does not
+    #       account for the Forem's administrators.
+    # @note For those considering extending this, be very mindful of
+    #       SQL injection.
+    class WeightedQueryStrategy
+      # This constant defines the allowable relevance scoring methods.
+      #
+      # A scoring method should be a SQL fragement that produces a
+      # value between 0 and 1.  The closer the value is to 1, the more
+      # relevant the article is for the given user.  Note: the values
+      # are multiplicative.  Make sure to consider if you want a 0
+      # multiplier for your score.  Aspirationally, you may want to
+      # think of the relevance_score as the range (0,1].  That is
+      # greater than 0 and less than or equal to 1.
+      #
+      # In addition, as part of initialization, the caller can
+      # configure each of the scoring methods :cases and :fallback.
+      #
+      # Each scoring method has the following keys:
+      #
+      # - clause: The SQL clause statement; note: there exists a
+      #           coupling between the clause and the SQL fragments
+      #           that join the various tables.  Also, under no
+      #           circumstances should you allow any user value for
+      #           this, as it is not something we can sanitize.
+      #
+      # - cases: An Array of Arrays, the first value is what matches
+      #          the clause, the second value is the multiplicative
+      #          factor.
+      #
+      # - fallback: When no case is matched use this factor.
+      #
+      # - requires_user: Does this scoring method require a given
+      #                  user.  If not, don't use it if we don't have
+      #                  a nil user.
+      #
+      # - group_by: An SQL fragment that ensures a valid postgres
+      #             statement in older versions of postgres.  See
+      #             https://github.com/forem/forem/pull/15240#discussion_r750392321
+      #             for further sleuthing details.  When you reference
+      #             a field in the clause, you likely need to include
+      #             a corresponding :group_by attribute.
+      #
+      # @note The group by clause appears necessary for postgres
+      #       versions and Heroku configurations of current (as of
+      #       <2021-11-16 Tue>) DEV.to installations.
+      SCORING_METHOD_CONFIGURATIONS = {
```

Comment:
I just come to realize just how amazing this implementation is :heart_eyes:  it's so easy to expand and adjust this! I feel like it could be isolated into its own file so changes to it can be more easily tracked on version control.

Reply:
Thanks.  That's been my goal.  I've got some tiny reworking to consider (unless Jamie or a Postgres-er has some rethinking of the LEFT OUTER JOINS) so that I don't include joins when I am not using a scoring method.

Reply:
@maestromac I went ahead and atomized it even further.  This should help us not add unnecessary joins when we don't select a scoring method.

This might also be a way to make a concession about not yet using some of the more expensive joins that @jgaskins identified.